optaplanner-core/src/main/java/org/optaplanner/core/impl/domain/valuerange/buildin/temporal/TemporalValueRange.java

```diff
@@ -0,0 +1,144 @@
+     * @param incrementUnit depends on the supported units of {from} and {to}
+     */
+    public TemporalValueRange(Temporal from, Temporal to, long incrementAmount, TemporalUnit incrementUnit) {
+        this.from = from;
+        this.to = to;
+        this.incrementAmount = incrementAmount;
+        this.incrementUnit = incrementUnit;
+
+        if (from == null || to == null || incrementUnit == null) {
+            throw new IllegalArgumentException("The " + getClass().getSimpleName()
+                    + " must have from (" + from + "),  to (" + to + "), incrementUnit (" + incrementUnit + ") which are not null.");
+        }
+
+        if (incrementAmount <= 0) {
+            throw new IllegalArgumentException("The " + getClass().getSimpleName()
+                    + " must have strictly positive incrementAmount (" + incrementAmount + ").");
+        }
+
+        if (!from.isSupported(incrementUnit) || !to.isSupported(incrementUnit)) {
+            throw new IllegalArgumentException("The " + getClass().getSimpleName()
+                    + " must have a incrementUnit which is supported by from/to (" + incrementUnit + ").");
+        }
+
+        if (from.until(to, incrementUnit) < 0) {
+            throw new IllegalArgumentException("The " + getClass().getSimpleName()
+                    + " cannot have a from (" + from + ") which is strictly higher than its to (" + to + ").");
+        }
+
+
+    }
+
+    private long getDuration() {
+        return from.until(to, incrementUnit);
+    }
+
+    @Override
+    public long getSize() {
+        return (getDuration() / incrementAmount);
+    }
+
+    @Override
+    public Temporal get(long index) {
+        if (index >= getSize() || index < 0) {
+            throw new IndexOutOfBoundsException();
+        }
+        return from.plus(index * incrementAmount, incrementUnit);
+    }
+
+    @Override
+    public Iterator<Temporal> createOriginalIterator() {
+        return new OriginalTemporalValueRangeIterator();
+    }
+
+    private class OriginalTemporalValueRangeIterator extends ValueRangeIterator<Temporal> {
+
+        private Temporal upcoming = from;
+
+        @Override
+        public boolean hasNext() {
+            return upcoming.until(to, incrementUnit) >= incrementAmount;
+        }
+
+        @Override
+        public Temporal next() {
+            if (upcoming.until(to, incrementUnit) < 0) {
+                throw new NoSuchElementException();
+            }
+
+            Temporal next = upcoming;
+            upcoming = upcoming.plus(incrementAmount, incrementUnit);
+            return next;
+        }
+    }
+
+    @Override
+    public boolean contains(Temporal value) {
+        if (value == null || !value.isSupported(incrementUnit)) {return false;}
+
+        // long delta = ...
+        // return from.until(value, incrementUnit) >= 0 && to.until(value, incrementUnit) < 0 && delta == 0;
+
+        for (long i = 0; i < getSize(); i++) {
```

Comment:
There's a scaling issue here (BigO notation is bad). I 'll take a look at a better scaling implementation. But the current implementation is a good start of course :)